### PR TITLE
[CALCITE-2000] UNNEST a collection that has a field with nested data generates an Exception

### PR DESCRIPTION
**Pull request by rubenada**
See details in Jira: [CALCITE-2000](https://issues.apache.org/jira/browse/CALCITE-2000)

### CODE REVIEW

**Comment by rubenada** [comment]
@vlsi any chance you can get a look at this one, please? I'd like to get this fixed for 1.27.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 1]
```diff
@@ -2792,27 +2792,28 @@ public static Collection multisetUnionAll(Collection collection1,
    * Function that, given a certain List containing single-item structs (i.e. arrays / lists with
    * a single item), builds an Enumerable that returns those single items inside the structs.
    */
-  public static Function1<Object, Enumerable<Comparable>> flatList() {
+  public static Function1<Object, Enumerable<Object>> flatList() {
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2795]: Are `null` values acceptable for the resulting function?
In other words, is `inputObject == null` valid or not?

If it is valid, then the signature should better be

```java
What do you think of the following signature?

```suggestion
  public static <T> Function1<T, Enumerable<T>> flatList() {
```

If null is not valid, then cast from `inputObject` to `List` should indeed be like `(List<Object>) ...` which effectively means `the list of non-nullable Object`

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2795]: {{SqlFunction#flatList}} was introduced in 1.24 via CALCITE-4063, and it is a counter-part of {{flatProduct}} for a very particular case. So, following {{flatProduct}}'s logic, inputObject cannot be null, so I guess {{flatList}} should look like:
```
  public static Function1<Object, Enumerable<Object>> flatList() {
    return inputObject -> {
      //noinspection unchecked
      final List<Object> list = (List<Object>) inputObject;
      final Enumerator<Object> enumerator = Linq4j.enumerator(list);
      ...
```
do you agree?

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2795]: Actually, I think we can do even better / simpler:
```
  public static Function1<List<Object>, Enumerable<Object>> flatList() {
    return inputObject -> {
      final Enumerator<Object> enumerator = Linq4j.enumerator(inputObject);
      ...
```

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Yes, that would be better indeed.
PS. `new Abstract ...` could probably be replaced with `Linq4j.asEnumerable(inputList).select(v -> ...)`

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 2]
```diff
@@ -2792,27 +2792,26 @@ public static Collection multisetUnionAll(Collection collection1,
...
               final Object element = enumerator.current();
-              final Comparable comparable;
+              final Object current;
               if (element.getClass().isArray()) {
-                comparable = (Comparable) ((Object[]) element)[0];
+                current = ((Object[]) element)[0];
               } else {
-                comparable = (Comparable) ((List) element).get(0);
+                current = ((List) element).get(0);
               }
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2813]: What do you think of reusing `org.apache.calcite.runtime.SqlFunctions#structAccess` here?

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Done.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 1 above)
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Done. Definitely better, thanks for the tip!

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 3]
```diff
@@ -2792,40 +2792,8 @@ public static Collection multisetUnionAll(Collection collection1,
...
-
-            @Override public void close() {
-              enumerator.close();
-            }
-          };
-        }
-      };
-    };
+  public static Function1<List<Object>, Enumerable<Object>> flatList() {
+    return inputList -> Linq4j.asEnumerable(inputList).select(v -> structAccess(v, 0, null));
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2796]: Great. Now it is much easier to understand what this call does :)

**Comment by rubenada** [comment]
@vlsi do you have any other comment about this PR? Do you think it is in a good shape for merge?

**Comment by vlsi** [comment]
I think it is good to go

**Comment by rubenada** [comment]
Thanks for your help @vlsi !

**Pull request merged**